Minimal NestedTensor

cpuhrsch opened the pull request:
This PR adds a minimal version of a NestedTensor. It introduces the general harness future development can be built around.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -11235,3 +11236,11 @@
   variants: function
   python_module: nn
+
+- func: nested_tensor_constructor(Tensor[] list, ScalarType dtype, Device device, bool pin_memory, bool channels_last) -> Tensor
```
are the tensor options to handle the case when the input list is empty? Shouldn't these be kwarg only?

cpuhrsch replied:
They can be. This is wrapped within Python, so it doesn't quite matter that much. I can fix this so that we can do away with the Python wrapper and use the native function only.

ezyang replied:
If it's wrapped in Python then give this function a leading underscore.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -11235,3 +11236,11 @@
+- func: nested_tensor_constructor(Tensor[] list, ScalarType dtype, Device device, bool pin_memory, bool channels_last) -> Tensor
+  variants: function
+
+- func: is_nt_impl(Tensor nt) -> bool
```
This looks a bit questionable

cpuhrsch replied:
I wish there was a better way. We don't seem to have a generic function that checks whether a torch.Tensor is backed by a particular TensorImpl subclass. Writing ```is_nested_tensor_impl``` seemed verbose, but I'm happy to change the name.

ezyang replied:
So, if we were to do this completely analogously to sparse, there wouldn't be a native_functions.yaml entry, and various implementations of is_nested would be directly implemented in relevant areas. TBH, the way these are currently added is kind of crappy, but if it's crappy in the same way as is_sparse/is_sparse_csr it will be easier for us to fix it later.

ezyang commented on torch/nestedtensor/nested/creation.py:
```diff
@@ -0,0 +1,33 @@
+import torch
```
Why not just `torch.nested` (in analogy to `torch.sparse`?)

cpuhrsch replied:
Works for me.

ezyang commented on torch/nestedtensor/nested/nested.py:
```diff
@@ -0,0 +1,133 @@
+        return cls.__dict__[name]
+
+
+class NestedTensor(metaclass=NestedTensorMeta):
```
I hate to be the naming, but let's be the naming guy. The internal representation here matches those being used by the preexisting faster transformer libraries. I am wondering: do they have other names they refer to this representation as (or do they all call it nested tensor)? In particular, "ragged tensor" is the TF nomenclature for what I believe is the same data structure (although, if I am reading https://github.com/NVIDIA/FasterTransformer/blob/main/docs/encoder_guide.md#standard-bert-and-effective-fastertransformer correctly, none of the row encodings in ragged tensor match the id offset encoding; the trt sequence length offset appears to match `row_splits` ? https://www.tensorflow.org/guide/ragged_tensor#raggedtensor_encoding )

Even if we don't decide to change any of the names here, it will be very helpful for users coming from other settings (e.g., preexisting FasterTransformers users or TF users) to know what NestedTensor is referred to in other contexts. But I also wonder: NestedTensor was originally coined when we were imagining arbitrary amounts of nesting, and it has sense been simplified to the form it is here; maybe nested tensor implies more "nesting" than you should actually expect?

cpuhrsch replied:
Any name you want is the name we will use.

ezyang commented on torch/nested/creation.py:
```diff
@@ -0,0 +1,33 @@
+
+def nested_tensor(
+    data,
+    dtype=None,
```
kwarg only plz

ezyang commented on torch/nested/creation.py:
```diff
@@ -0,0 +1,33 @@
+    device=None,
+    requires_grad=False,
+    pin_memory=False,
+    channels_last=False,
```
this is not symmetric with how factory functions work. I'd prefer if this API lined up with our preexisting factory functions

ezyang replied:
Also it's kind of weird that you have this flag but it has to be False. Why not just delete it for now (and at it back later when memory format support comes online?)

cpuhrsch replied:
Updating it right now. Will orient myself on functions such as ones etc.

ezyang commented on torch/nested/_nestedtensor.py:
```diff
@@ -0,0 +1,118 @@
+        torch.NestedTensor(result)
+        if torch.is_tensor(result) and torch.is_nt_impl(result)
+        else result
+    )
```
Consider using tree_map here?

cpuhrsch replied:
I did, but remember it wasn't working. Will revisit.

ezyang commented on torch/nested/_nestedtensor.py:
```diff
@@ -0,0 +1,118 @@
+    )
+
+
+class NestedTensor:
```
Need more docs...

cpuhrsch replied:
Can I add docs in a follow up?

ezyang replied:
Also need an explanation of why we are wrapping NestedTensor in this way

cpuhrsch replied:
Because it doesn't have a size. So, something as simple as printing doesn't work. We should migrate this to torch_dispatch as a follow-up.

ezyang replied:
I think the way of fixing printing which is more in line with how, e.g., sparse is currently done, is to just override the tensor printer to check for nested tensor (so you'll need to bind an `is_nested_tensor` check somehow; current prevailing style is to do it as a method but TBH for experimental stuff it's better to cook it as a function first). You delegate size call to the inner implementation so removing the wrapper is no worse...

cpuhrsch replied:
Yes, now that we're in tree that is feasible. Out of tree that is impossible. That's why this is written this way.

ezyang replied:
OK I filed an issue for printing specifically at https://github.com/pytorch/pytorch/issues/73506 I'm not sure what you want to do about serialization though: because it's a C++ struct difference the logical interposition point would be in C++, but pickling is a purely Python side idea (TorchScript C++ pickler/unpickler notwithstanding), so you've gotta put the logic somewhere.

ezyang commented on torch/nested/_nestedtensor.py:
```diff
@@ -0,0 +1,118 @@
+        to reach a Tensor constiuent.
+        """
+        # This NT only supports nesting of 1.
+        return 1
```
I'd axe this for now

ezyang commented:
@swolchok if you could give this a look over and see if this is enough for your purposes, that would be a big help

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,92 @@
+inline at::native::NestedTensorImpl* get_nested_tensor_impl(
+    const at::Tensor tensor) {
+  if (!is_nested_tensor_impl(tensor)) {
+    throw std::runtime_error("Function requires NestedTensorImpl");
```
use the `TORCH_CHECK` macro? (or maybe internal assert, if the nested tensor precondition is supposed to be satisfied by caller)

cpuhrsch commented:
@swolchok - I know you've been landing various performance optimizations in pytorch/nestedtensor and will port them to core in follow up PRs. We can then eventually have pytorch/nestedtensor depend on torch.NestedTensor.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,91 @@
+  std::vector<Tensor> sizes;
+  std::vector<Tensor> flat_tensors;
+  for (size_t i = 0; i < list.size(); i++) {
+    flat_tensors.push_back(list[i].reshape(-1).contiguous());
```
it's not clear to me why the contiguous() call is necessary; shouldn't cat be able to deal with non-contiguous inputs?

cpuhrsch replied:
NestedTensor currently has no concept of strides. This will be done in follow up PRs.

swolchok replied:
maybe leave a todo to remove it?

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,91 @@
+    bool pin_memory,
+    bool channels_last) {
+  std::vector<Tensor> sizes;
+  std::vector<Tensor> flat_tensors;
```
Are there any invariants on the dims of the input tensors in TensorList?

cpuhrsch replied:
All Tensors must have the same dimension. I can add a check for this.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,91 @@
+  std::vector<Tensor> flat_tensors;
+  for (size_t i = 0; i < list.size(); i++) {
+    flat_tensors.push_back(list[i].reshape(-1).contiguous());
+    sizes.push_back(at::tensor(c10::IntArrayRef(list[i].sizes())));
```
This doesn't look like the FT offsets representation (https://github.com/NVIDIA/FasterTransformer/blob/main/docs/encoder_guide.md#standard-bert-and-effective-fastertransformer) what's going on here?

cpuhrsch replied:
You can easily convert into offsets or id based representations. This is more compact, which is a choice. See [to_padded_tensor](https://github.com/pytorch/nestedtensor/blob/master/nestedtensor/csrc/masking.cpp#L571) as an example. We might later on cache the offsets as an optimization, but it doesn't always make sense. In particular if the innermost dimension is irregular storing the offsets as a flat list might not be worth it.

cpuhrsch commented:
I'm currently trying to switch to torch_dispatch, because it's better. I'm running into the following error when trying to call unbind on it.

```
         (source) /dev/shm/cpuhrsch/pytorch - git:addnestedtensor1]$ python /fsx/users/cpuhrsch/conda/envs/source/bin/py.test test/test_nestedtensor.py -k "test_repr_string" -s                                             [32/1988]
======================================================================================================== test session starts =========================================================================================================
         linux -- Python 3.9.5, pytest-6.2.4, py-1.10.0, pluggy-0.13.1                                                                                                                                                       [30/1988]
rootdir: /dev/shm/cpuhrsch/pytorch, configfile: pytest.ini
plugins: hypothesis-6.14.1
collected 11 items / 10 deselected / 1 selected
                                                                                                                                                                                                                             [26/1988]
test/test_nestedtensor.py func:  OpOverloadPacket(op='aten.unbind')
args:  <class 'torch.nested._nestedtensor.NestedTensor'>
impl_args:  <class 'torch.Tensor'>
func:  OpOverloadPacket(op='aten.unbind')
args:  <class 'torch.nested._nestedtensor.NestedTensor'>
impl_args:  <class 'torch.Tensor'>
func:  OpOverloadPacket(op='aten.unbind')
args:  <class 'torch.nested._nestedtensor.NestedTensor'>
impl_args:  <class 'torch.Tensor'>
func:  OpOverloadPacket(op='aten.unbind')
args:  <class 'torch.nested._nestedtensor.NestedTensor'>
F

============================================================================================================== FAILURES ==============================================================================================================
_________________________________________________________________________________________________ TestNestedTensor.test_repr_string _________________________________________________________________________________________[11/1988$
          (most recent call last):                                                                                                                                                                                           [10/1988]
  File "/dev/shm/cpuhrsch/pytorch/test/test_nestedtensor.py", line 170, in test_repr_string
    self.assertEqual(str(a), expected)
  File "/dev/shm/cpuhrsch/pytorch/torch/nested/_nestedtensor.py", line 107, in __str__
    return "nested_tensor(" + _str(self) + ")"
  File "/dev/shm/cpuhrsch/pytorch/torch/nested/_nestedtensor.py", line 94, in _str
    strs = list(map(str, x.unbind()))
  File "/dev/shm/cpuhrsch/pytorch/torch/nested/_nestedtensor.py", line 107, in __str__
    return "nested_tensor(" + _str(self) + ")"
  File "/dev/shm/cpuhrsch/pytorch/torch/nested/_nestedtensor.py", line 94, in _str
    strs = list(map(str, x.unbind()))
  File "/dev/shm/cpuhrsch/pytorch/torch/_tensor.py", line 1142, in __torch_function__
    ret = func(*args, **kwargs)
  File "/dev/shm/cpuhrsch/pytorch/torch/nested/_nestedtensor.py", line 116, in __torch_dispatch__
    impl_args, impl_kwargs = _filter_impl(args, kwargs)
  File "/dev/shm/cpuhrsch/pytorch/torch/nested/_nestedtensor.py", line 12, in _filter_impl
    impl_args.append(a._impl)
AttributeError: 'NestedTensor' object has no attribute '_impl'
```

Also _make_tensor_subclass *requires* a size, which NestedTensor can't provide. Right now I'm just stuffing it with 1s and of length matching the dimension of the Tensor.

I'm not sure how to proceed here in using the Tensor subclassing mechanisms.

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,122 @@
+          c10::DispatchKeySet({DispatchKey::NestedTensor}),
+          buffer.dtype(),
+          buffer.device()),
+      _buffer(buffer),
+      _nested_size_tensor(nested_size_tensor) {
```
std::move

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,122 @@
+at::Tensor wrap_buffer(at::Tensor buffer, at::Tensor nested_size_tensor) {
+  TORCH_CHECK(buffer.is_contiguous(), "Given buffer must be contiguous.");
+  return at::detail::make_tensor<NestedTensorImpl>(
+      std::move(buffer), nested_size_tensor);
```
can we not move nested_size_tensor too?

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,122 @@
+  }
+  auto esizes_chunks = esizes.unbind(0);
+  std::vector<int64_t> splits;
+  for (int64_t i = 0; i < esizes_chunks.size(); i++) {
```
use c10::irange

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,122 @@
+  }
+  auto buffer_chunks = at::split_with_sizes(buffer, IntArrayRef(splits));
+  for (int64_t i = 0; i < buffer_chunks.size(); i++) {
+    auto esize_chunk = esizes_chunks[i];
```
`const auto&`

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,122 @@
+  auto buffer_chunks = at::split_with_sizes(buffer, IntArrayRef(splits));
+  for (int64_t i = 0; i < buffer_chunks.size(); i++) {
+    auto esize_chunk = esizes_chunks[i];
+    std::vector<int64_t> esize_vector(
+        esize_chunk.data_ptr<int64_t>(),
+        esize_chunk.data_ptr<int64_t>() + esize_chunk.numel());
+    result_tensors.push_back(buffer_chunks[i].view(IntArrayRef(esize_vector)));
```
no need for the vector; make an IntArrayRef directly

cpuhrsch replied:
IntArrayRef can't be constructed from a Tensor.

```
    ../aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp: In function ‘std::vector<at::Tensor> at::native::NestedTensor_unbind(const at::Tensor&, int64_t)
’:
    ../aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:54:75: error: no matching function for call to ‘c10::ArrayRef<long int>::ArrayRef(const at::Tenso
r&)’
         result_tensors.push_back(buffer_chunks[i].view(IntArrayRef(esize_chunk)));                                                                          ^
```

swolchok replied:
you can construct one using the `ArrayRef(const T* begin, const T* end)` constructor just as you are constructing the vector here:

```
  IntArrayRef esize_arrayref(
        esize_chunk.data_ptr<int64_t>(),
        esize_chunk.data_ptr<int64_t>() + esize_chunk.numel());
```

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,92 @@
+namespace at {
+namespace native {
+
+bool is_nested_tensor_impl(at::Tensor tensor) {
```
const at::Tensor&

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,92 @@
+};
+
+inline at::native::NestedTensorImpl* get_nested_tensor_impl(
+    const at::Tensor tensor) {
```
const at::Tensor&

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,90 @@
+    TORCH_CHECK(
+        false,
+        "Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor");
+    std::vector<int64_t> sizes;
+    return IntArrayRef(sizes);
```
this will just produce a dangling IntArrayRef. I would expect compilers to be able to handle omitting it, but I'm guessing some compiler we support can't handle it, so I would suggest `return IntArrayRef()` instead

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,90 @@
+    TORCH_CHECK(
+        false,
+        "Internal error: NestedTensorImpl doesn't support strides. Please file an issue on https://github.com/pytorch/nestedtensor");
+    std::vector<int64_t> strides;
+    return IntArrayRef(strides);
```
same

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,90 @@
+      tensor.unsafeGetTensorImpl());
+}
+
+inline at::Tensor get_buffer(const at::Tensor& tensor) {
```
return const reference ( I think I recently landed a commit that fixes this and adds a variant that returns a non-const reference given non-const tensor reference as well)

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,90 @@
+  return get_nested_tensor_impl(tensor)->get_buffer();
+}
+
+inline const at::Tensor get_nested_size_tensor(const at::Tensor& tensor) {
```
same

swolchok commented on torch/nested/_nestedtensor.py:
````diff
@@ -0,0 +1,100 @@
+        """
+        return self._impl.stride()
+
+    def size(self):
+        """
+        The size of ```self``` NestedTensor.
+        """
+        return self._impl.size()
````
why does this exist if it doesn't have a size?

cpuhrsch replied:
To throw an exception

swolchok replied:
if that's expected to always be the case, perhaps that should be noted in the docstring rather than being a surprise?

cpuhrsch replied:
Sure, let me update

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,109 @@
+      dim,
+      " instead.");
+  auto esizes = get_nested_size_tensor(self);
+  auto buffer = get_buffer(self);
```
`const auto&`
nit: move down to point of use (or don't make it a local with a name at all, which solves the `const auto&` thing to) to reduce cognitive load

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,109 @@
+    splits.push_back(esizes_chunks[i].prod().item<int64_t>());
+  }
+  auto buffer_chunks = at::split_with_sizes(buffer, IntArrayRef(splits));
+  for (int64_t i = 0; i < buffer_chunks.size(); i++) {
```
nit: ideally use irange but I'm not going to block over it

swolchok commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,109 @@
+  }
+  std::vector<Tensor> sizes;
+  std::vector<Tensor> flat_tensors;
+  for (size_t i = 0; i < list.size(); i++) {
```
irange?

swolchok reviewed (commented):
no serious complaints. @ezyang what do you think?

ezyang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -79,6 +79,9 @@ file(GLOB miopen_cpp "miopen/*.cpp")
 file(GLOB mkldnn_cpp "mkldnn/*.cpp")
 
+file(GLOB nestedtensor_h "native/nestedtensor/*.h")
+file(GLOB nestedtensor_cpp "native/nestedtensor/*.cpp")
```
I hope you're prepared to update all the bzl build files lol, we definitely do not recursively glob under native 📦

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -11240,3 +11241,6 @@
   variants: function
   python_module: nn
+
+- func: _nested_tensor(Tensor[] list, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Still no kwarg only after the list? Well it doesn't matter too much here...

BTW, in the minimal formulation, it might be helpful to also have a constructor that takes the compressed Tensor representation directly, instead of having to copy things in. But @swolchok can always add that if he needs it, I suppose.

ezyang replied:
OK I filed an issue for this at https://github.com/pytorch/pytorch/issues/73505

However, to me, this is a pretty good indication that you should have exposed the buffer+nested size constructor (with no TensorOptions arguments) rather than this constructor. You can then provide this implementation in Python.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+    at::Tensor buffer,
+    at::Tensor nested_size_tensor)
+    : TensorImpl(
+          c10::DispatchKeySet({DispatchKey::NestedTensor}),
```
If you want a CPU nested tensor to properly report `is_cpu` you also need to include the CPU/CUDA dispatch key as necessary. (Or I guess you can wait until @bdhirsh lands https://github.com/pytorch/pytorch/pull/72827 and then cartesian product nested with cpu/cuda, which is probably a good idea anyway)

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+      std::move(buffer), std::move(nested_size_tensor));
+}
+
+std::vector<at::Tensor> NestedTensor_unbind(
```
NB: this only works for CPU

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+    return result_tensors;
+  }
+  auto esizes_chunks = esizes.unbind(0);
+  std::vector<int64_t> splits;
```
you can reserve on this

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+      "got dimension ",
+      dim,
+      " instead.");
+  auto esizes = get_nested_size_tensor(self);
```
const auto& instead, avoid refcount bump

ezyang replied:
Where does the name esize come from? I actually don't mind it as an abbreviated name, it's just jarring that it's "officially" called "nested_size_tensor"

cpuhrsch replied:
We can come up with a better name in a follow-up PR.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+  std::vector<int64_t> splits;
+  for (const auto i : c10::irange(esizes_chunks.size())) {
+    splits.push_back(esizes_chunks[i].prod().item<int64_t>());
+  }
```
This algorithm here feels like there ought to be a kernel that can do it directly, rather than unbinding it and doing it manually yourself. Like, shouldn't this be `torch.prod(esizes, dim=1)` (and maybe a dtype conversion to long) and then you can just view the memory as an IntArrayRef directly without having to allocate the vector.

ezyang replied:
esize_chunks is used again below but the use there also feels unnecessary

cpuhrsch replied:
Let's clean this up as part of profile-guided optimization based on use cases.

cpuhrsch replied:
For now I want to aim for correctness and development speed. There is a lot we can do here, including finding a more efficient data structure to store these numbers to begin with.

ezyang replied:
Yeah, I'm heuristic'ing here "ops involving small tensors are antipattern". It's OK to defer to swolchok here I guess.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+    splits.push_back(esizes_chunks[i].prod().item<int64_t>());
+  }
+  auto buffer_chunks =
+      at::split_with_sizes(get_buffer(self), IntArrayRef(splits));
```
IntArrayRef here is unnecessary

cpuhrsch replied:
Removed

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+          buffer.dtype(),
+          buffer.device()),
+      _buffer(std::move(buffer)),
+      _nested_size_tensor(std::move(nested_size_tensor)) {
```
any requirements on nested_size_tensor? Looks like it has to be contiguous to me..

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:
```diff
@@ -0,0 +1,110 @@
+
+  return wrap_buffer(
+      at::native::cat(at::TensorList(flat_tensors)).to(options),
+      at::native::stack(at::TensorList(sizes)));
```
TensorList conversion here is unnecessary

ezyang replied:
Is the device of the values and the esizes supposed to match? esizes seems pretty necessary to live on CUDA for hypothetical CUDA kernels?

cpuhrsch replied:
Not necessarily. Size manipulation (such as via view operations or offset conversions) can be faster on the CPU than constant kernel launches. Let's do profile guided optimization and decide on this once we have models and ops to optimize.

ezyang replied:
If CUDA is out of scope we can kick this down the road. But for CUDA, PGO is cool but it is not a substitute for deliberate design of the core data structures in the first place. If there are kernels where result values depend on the actual sizes, then I would expect CUDA sizes to always win as otherwise you will have to host-to-device transfer every time you hit a kernel like this.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,84 @@
+  explicit NestedTensorImpl(at::Tensor buffer, at::Tensor nested_size_tensor);
+
+  int64_t dim() const override {
+    return _nested_size_tensor.dim() > 0 ? 1 + _nested_size_tensor.size(1) : 1;
```
instead of overriding this, can we simply just allocate size_ field with the appropriate number of entries (even if those entries contain garbage?)

cpuhrsch replied:
We can. Let's do that once we find this operation is slow in the context of a set of operations and it shows up in a profile. For now I want to aim for correctness.

ezyang replied:
This is about correctness though? Scott has devirtualized many of the size related methods on TensorImpl (a development I wholeheartedly support), and so custom overrides like this will work differently inside fbcode. If we can dispense with the override entirely we can also guarantee identical behavior in OSS and fbcode.

BTW, I've kind of gotten to the point where I think nested tensors should have an actual size, and simply MAX_INT for any ragged dimensions. I think this will work pretty well, since you cannot actually exceed MAX_INT entries in the ragged representation. MAX_INT is also stupidly large and you will get an error if you try to use it for another allocation.

cpuhrsch replied:
Sure we can add that, but it would be great if calling .size() on NestedTensorImpl were to actually throw, because I've been bitten in the past by calling into composite operators or other unsupported operations unknowingly and getting very weird errors (for example matmul will call size). Any NestedTensor kernel that wants to use size() unconditionally is doing something wrong (note that there is an interpretation for size(i) if i is regular). I've worked around this in pytorch/nestedtensor by wrapping call devirtualized method calls into e.g. ```get_size```, ```get_dtype``` etc.

ezyang commented on aten/src/ATen/native/nestedtensor/NestedTensorImpl.h:
```diff
@@ -0,0 +1,84 @@
+
+ private:
+  at::Tensor _buffer;
+  const at::Tensor _nested_size_tensor;
```
Conventionally, fields in C++ objects are named `foo_`, even if they are private (and it is assumed that members are private, always.)

ezyang commented on test/test_nestedtensor.py:
```diff
@@ -0,0 +1,173 @@
+
+
+class TestNestedTensor(TestCase):
+    @torch.inference_mode()
```
Is this required?

cpuhrsch replied:
Yes, it is. NestedTensor does not support autograd. If omitted we run into strange errors (because composites that can't deal with NestedTensor are being called). We can register a fallback under AutogradNestedTensor that throws an exception if NestedTensor is called outside this context. Let's address this in a follow-up PR.

ezyang replied:
OK, this is fine to follow up, but we definitely should fix this; the system should be capable of handling this case correctly.

ezyang commented on test/test_nestedtensor.py:
```diff
@@ -0,0 +1,173 @@
+                torch.tensor([]),
+                torch.tensor([]),
+                torch.tensor([]),
+            )
```
If you use subtests or parametrize, that will probably make for more pleasant test debugging if one particular test fails (cuz you can tell which subtest you were in without having to pdb)

ezyang commented on test/test_nestedtensor.py:
```diff
@@ -0,0 +1,173 @@
+
+                a = torch.randn((2, 3)).add_(1)
+                nt = nested_tensor([a])
+                self.assertEqual(a, unbind_fn(nt, 0)[0])
```
c, d, e are dead, I'm guessing cuz you want to port more tests here in the future?

cpuhrsch replied:
Will remove.

ezyang commented:
Structurally, this all seems fine, the comments are about tightening loose bolts and hatches for core entry.

ezyang commented:
I pushed a few comment only commits for some of my CR comments that hadn't been incorporated into the codebase as TODOs in a durable way

albanD commented on docs/source/conf.py:
```diff
@@ -277,6 +277,7 @@
     "torch.quantization",
     "torch.quantization.fx",
     "torch.sparse",
+    "torch.nested",
```
We should not add undocumented public namespaces please.
As mentioned in the comment above. You should NOT add anything new to this list!

cpuhrsch replied:
There is documentation here actually, checkout the preview. https://docs-preview.pytorch.org/72881/index.html

albanD replied:
Perfect, so this line can be removed :)

albanD replied:
Thanks for the update! :)

cpuhrsch committed:
Minimal NestedTensor (#72881)

Summary:
This PR adds a minimal version of a NestedTensor. It introduces the general harness future development can be built around.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/72881

Reviewed By: albanD

Differential Revision: D34259177

Pulled By: cpuhrsch

fbshipit-source-id: 2ada74fadf786d89d00ff1973ab4126abbc1f3e3
